File: research/experiments/sweep/runs/opt-small-csharp-sonnet45/Optimization/VecOpsTests.cs
namespace Optimization;

public class VecOpsTests
{
    [Fact]
    public void Dot_BasicCase()
    {
        Assert.Equal(32.0, VecOps.Dot([1, 2, 3], [4, 5, 6]));
    }

    [Fact]
    public void Dot_ZeroVectors()
    {
        Assert.Equal(0.0, VecOps.Dot([0, 0], [1, 1]));
    }

    [Fact]
    public void Norm_ThreeFour()
    {
        Assert.Equal(5.0, VecOps.Norm([3, 4]));
    }

    [Fact]
    public void Norm_Zero()
    {
        Assert.Equal(0.0, VecOps.Norm([0, 0, 0]));
    }

    [Fact]
    public void NormInf_Mixed()
    {
        Assert.Equal(3.0, VecOps.NormInf([1, -3, 2]));
    }

    [Fact]
    public void NormInf_Zero()
    {
        Assert.Equal(0.0, VecOps.NormInf([0, 0]));
    }

    [Fact]
    public void Scale_Basic()
    {
        Assert.Equal([3.0, 6.0], VecOps.Scale([1, 2], 3));
    }

    [Fact]
    public void Scale_Zero()
    {
        Assert.Equal([0.0, 0.0], VecOps.Scale([1, 2], 0));
    }

    [Fact]
    public void Add_Basic()
    {
        Assert.Equal([4.0, 6.0], VecOps.Add([1, 2], [3, 4]));
    }

    [Fact]
    public void Sub_Basic()
    {
        Assert.Equal([2.0, 2.0], VecOps.Sub([3, 4], [1, 2]));
    }

    [Fact]
    public void Negate_Basic()
    {
        Assert.Equal([-1.0, 2.0], VecOps.Negate([1, -2]));
    }

    [Fact]
    public void Clone_CreatesNewArray()
    {
        double[] original = [1, 2];
        double[] cloned = VecOps.Clone(original);
        Assert.Equal([1.0, 2.0], cloned);

        // Verify it's a distinct array
        cloned[0] = 99;
        Assert.Equal(1.0, original[0]);
    }

    [Fact]
    public void Zeros_CreateZeroVector()
    {
        Assert.Equal([0.0, 0.0, 0.0], VecOps.Zeros(3));
    }

    [Fact]
    public void AddScaled_Basic()
    {
        Assert.Equal([7.0, 10.0], VecOps.AddScaled([1, 2], [3, 4], 2));
    }

    [Fact]
    public void Add_DoesNotMutateInputs()
    {
        double[] a = [1, 2];
        double[] b = [3, 4];
        VecOps.Add(a, b);
        Assert.Equal([1.0, 2.0], a);
        Assert.Equal([3.0, 4.0], b);
    }

    [Fact]
    public void Scale_DoesNotMutateInput()
    {
        double[] v = [1, 2];
        VecOps.Scale(v, 3);
        Assert.Equal([1.0, 2.0], v);
    }
}

File: research/experiments/sweep/runs/opt-small-swift-sonnet45/Tests/OptimizationTests/VecOpsTests.swift
import XCTest
@testable import Optimization

final class VecOpsTests: XCTestCase {

    // MARK: - dot

    func testDot() {
        XCTAssertEqual(VecOps.dot([1, 2, 3], [4, 5, 6]), 32, accuracy: 1e-10)
    }

    func testDotZero() {
        XCTAssertEqual(VecOps.dot([0, 0], [1, 1]), 0, accuracy: 1e-10)
    }

    // MARK: - norm

    func testNorm() {
        XCTAssertEqual(VecOps.norm([3, 4]), 5, accuracy: 1e-10)
    }

    func testNormZero() {
        XCTAssertEqual(VecOps.norm([0, 0, 0]), 0, accuracy: 1e-10)
    }

    // MARK: - normInf

    func testNormInf() {
        XCTAssertEqual(VecOps.normInf([1, -3, 2]), 3, accuracy: 1e-10)
    }

    func testNormInfZero() {
        XCTAssertEqual(VecOps.normInf([0, 0]), 0, accuracy: 1e-10)
    }

    // MARK: - scale

    func testScale() {
        XCTAssertEqual(VecOps.scale([1, 2], 3), [3, 6])
    }

    func testScaleZero() {
        XCTAssertEqual(VecOps.scale([1, 2], 0), [0, 0])
    }

    // MARK: - add

    func testAdd() {
        XCTAssertEqual(VecOps.add([1, 2], [3, 4]), [4, 6])
    }

    func testAddDoesNotMutate() {
        let a = [1.0, 2.0]
        let b = [3.0, 4.0]
        let aCopy = a
        let bCopy = b
        _ = VecOps.add(a, b)
        XCTAssertEqual(a, aCopy)
        XCTAssertEqual(b, bCopy)
    }

    // MARK: - sub

    func testSub() {
        XCTAssertEqual(VecOps.sub([3, 4], [1, 2]), [2, 2])
    }

    // MARK: - negate

    func testNegate() {
        XCTAssertEqual(VecOps.negate([1, -2]), [-1, 2])
    }

    // MARK: - clone

    func testClone() {
        let original = [1.0, 2.0]
        let cloned = VecOps.clone(original)
        XCTAssertEqual(cloned, [1, 2])
        // Verify it's a distinct array
        var mutableClone = cloned
        mutableClone[0] = 999
        XCTAssertEqual(original[0], 1.0)
    }

    // MARK: - zeros

    func testZeros() {
        XCTAssertEqual(VecOps.zeros(3), [0, 0, 0])
    }

    // MARK: - addScaled

    func testAddScaled() {
        XCTAssertEqual(VecOps.addScaled([1, 2], [3, 4], 2), [7, 10])
    }

    // MARK: - Purity checks

    func testScaleDoesNotMutate() {
        let v = [1.0, 2.0]
        let vCopy = v
        _ = VecOps.scale(v, 3)
        XCTAssertEqual(v, vCopy)
    }
}
